smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -0,0 +1,675 @@
+
+
+def main(
+    frequency: t.Union[int, float],
+    experiment_dir: pathlib.Path,
+    logger: logging.Logger,
+    observer: t.Optional[BaseObserver] = None,
+) -> int:
+    """Setup the monitoring entities and start the timer-based loop that
+    will poll for telemetry data
+
+    :param frequency: frequency (in seconds) of update loop
+    :type frequency: t.Union[int, float]
+    :param experiment_dir: the experiement directory to monitor for changes
+    :type experiment_dir: pathlib.Path
+    :param observer: (optional) a preconfigured Observer to inject
+    :type observer: t.Optional[BaseObserver]
+    """
+    manifest_relpath = pathlib.Path(TELMON_SUBDIR) / MANIFEST_FILENAME
+    manifest_path = experiment_dir / manifest_relpath
+    monitor_pattern = str(manifest_relpath)
+
+    logger.info(
+        f"Executing telemetry monitor with frequency: {frequency}s"
+        f", on target directory: {experiment_dir}"
+        f" matching pattern: {monitor_pattern}"
+    )
+
+    log_handler = LoggingEventHandler(logger)  # type: ignore
+    action_handler = ManifestEventHandler(monitor_pattern, logger)
+
+    if observer is None:
+        observer = Observer()
+
+    try:
+        if manifest_path.exists():
+            # a manifest may not exist depending on startup timing
+            action_handler.process_manifest(str(manifest_path))
+
+        observer.schedule(log_handler, experiment_dir, recursive=True)  # type:ignore
+        observer.schedule(action_handler, experiment_dir, recursive=True)  # type:ignore
+        observer.start()  # type: ignore
+
+        event_loop(observer, action_handler, frequency, logger)
+        return 0
+    except Exception as ex:
+        logger.error(ex)
+    finally:
+        if observer.is_alive():
+            observer.stop()  # type: ignore
+            observer.join()
+
+    return 1
+
+
+def handle_signal(signo: int, _frame: t.Optional[FrameType]) -> None:
+    """Helper function to ensure clean process termination"""
+    if not signo:
+        logger = logging.getLogger()
+        logger.warning("Received signal with no signo")
+
+
+def register_signal_handlers() -> None:
+    """Register a signal handling function for all termination events"""
+    for sig in SIGNALS:
+        signal.signal(sig, handle_signal)
+
+
+def get_parser() -> argparse.ArgumentParser:
+    """Instantiate a parser to process command line arguments"""
+    arg_parser = argparse.ArgumentParser(description="SmartSim Telemetry Monitor")
+    arg_parser.add_argument(
+        "-f",
+        type=str,
+        help="Frequency of telemetry updates",
+        required=True,
+    )
+    arg_parser.add_argument(
+        "-d",
+        type=str,
+        help="Experiment root directory",
+        required=True,
+    )
+    arg_parser.add_argument(
+        "-n",
+        type=int,
+        help="Automatically shutdown after a specific number of polling iterations",
+        default=0,
+        required=False,
+    )
```

Comment:
Sorry to be pedantic again, but can I request long args here as well 😅

Reply:
done!